src/main/java/org/jabref/logic/openoffice/oocsltext/CSLReferenceMarkManager.java

```diff
@@ -25,73 +32,158 @@
+
+            // Remove the old reference mark without removing the text
+            text.removeTextContent(oldContent);
+
+            // Update the text
+            cursor.setString(updatedText);
+
+            // Create a new reference mark with updated name
+            String updatedName = updateReferenceName(mark.getName(), newNumber);
+            XNamed newNamed = UnoRuntime.queryInterface(XNamed.class,
+                    factory.createInstance("com.sun.star.text.ReferenceMark"));
+            newNamed.setName(updatedName);
+            XTextContent newContent = UnoRuntime.queryInterface(XTextContent.class, newNamed);
+
+            // Attach the new reference mark to the cursor range
+            newContent.attach(cursor);
+
+            // Update our internal reference to the new text content and name
+            mark.updateTextContent(newContent);
+            mark.updateName(updatedName);
+            mark.setCitationNumber(newNumber);
+        }
+    }
+
+    private String updateReferenceName(String oldName, int newNumber) {
+        String[] parts = oldName.split(" ");
+        if (parts.length >= 2) {
+            parts[1] = "CID_" + newNumber;
+            return String.join(" ", parts);
+        }
+        return oldName;
+    }
+
+    private String updateCitationText(String currentText, int newNumber) {
+        Pattern pattern = Pattern.compile("(\\D*)(\\d+)(\\D*)");
+        Matcher matcher = pattern.matcher(currentText);
+        if (matcher.find()) {
+            String prefix = matcher.group(1);
+            String suffix = matcher.group(3);
+            return prefix + newNumber + suffix;
+        }
+        return currentText;
+    }
+
+    public int getCitationNumber(String citationKey) {
+        return citationKeyToNumber.computeIfAbsent(citationKey, k -> ++highestCitationNumber);
     }
 
     public void readExistingMarks() throws WrappedTargetException, NoSuchElementException {
+        marksByName.clear();
+        marksInOrder.clear();
+        citationKeyToNumber.clear();
+
         XReferenceMarksSupplier supplier = UnoRuntime.queryInterface(XReferenceMarksSupplier.class, document);
         XNameAccess marks = supplier.getReferenceMarks();
 
-        citationOrder.clear();
-        int citationCounter = 0;
-
         for (String name : marks.getElementNames()) {
-            Optional<ReferenceMark> referenceMark = ReferenceMark.of(name);
-            if (!referenceMark.isEmpty()) {
-                citationOrder.putIfAbsent(referenceMark.map(ReferenceMark::getCitationKey).get(), ++citationCounter);
+            if (name.startsWith("JABREF_")) {
                 XNamed named = UnoRuntime.queryInterface(XNamed.class, marks.getByName(name));
-                CSLReferenceMark mark = new CSLReferenceMark(named, referenceMark.get());
-                addMark(mark);
+                String[] parts = name.split(" ");
+                String citationKey = parts[0].substring(7);
```

Comment:
Please at least use some constants for the magic numbers and check the lengths of the parts array to prevent any exceptions